frontend/src/visualization/shared-ops/fetchDataFromService.tsx

```diff
@@ -0,0 +1,191 @@
+import Backdrop from '@material-ui/core/Backdrop';
+import CircularProgress from '@material-ui/core/CircularProgress';
+import Button from '@material-ui/core/Button';
+import CloseIcon from '@material-ui/icons/Close';
+import React, { useEffect, useRef, useState } from 'react';
+import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
+import {
+  CancellationToken,
+  CancellationTokenSource,
+} from '../../utils/CancellationToken';
+
+const useStyles = makeStyles((theme: Theme) =>
+  createStyles({
+    backdrop: {
+      zIndex: theme.zIndex.drawer + 1,
+      color: '#fff',
+    },
+    backdropContent: {
+      marginTop: theme.spacing(3),
+      display: 'flex',
+      flexDirection: 'column',
+      alignItems: 'center',
+    },
+    backdropCancel: {
+      marginTop: theme.spacing(3),
+    },
+    contentContainer: {
+      padding: theme.spacing(3),
+    },
+  })
+);
+
+type RenderFunction<T> = (t: T, update: () => void) => JSX.Element;
+
+type QueryFunction<
+  TArgs extends [...Array<unknown>, CancellationToken],
```

Comment:
This is very confusing: Here `TArgs` extends from `[...Array<unknown>, CancellationToken]`. On the other hand, later, `TArgs` always extends from  `Array<unknown>`. I would change this `TArgs` here in `QueryFunction` to another type since it makes the most sense to me that `TArgs` are the additional arguments WITHOUT cancellationToken.

Reply:
Done!